crates/cext/src/transpiler/passes/gate_direction.rs

```diff
@@ -0,0 +1,128 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2025
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use crate::pointers::{const_ptr_as_ref, mut_ptr_as_ref};
+
+use qiskit_circuit::circuit_data::CircuitData;
+use qiskit_circuit::converters::dag_to_circuit;
+use qiskit_circuit::dag_circuit::DAGCircuit;
+use qiskit_transpiler::passes::{check_direction_target, fix_direction_target};
+use qiskit_transpiler::target::Target;
+
+/// @ingroup QkTranspilerPasses
+/// Run the CheckGateDirection pass on a circuit.
+///
+/// The pass checks if the directions of two-qubit gates comply with the gate directions specified in a given target.
+///
+/// @param circuit A pointer to the circuit on which to run the CheckGateDirection pass.
+/// @param target A pointer to the target used to check gate directions.
+///
+/// @return bool - true iff the directions of all two-qubit gates in the circuit comply with specified target constraints
+///
+/// # Example
+///
+/// ```c
+///    QkTarget *target = qk_target_new(2);
+///    uint32_t qargs[3] = {0,1};
+///
+///    QkTargetEntry *cx_entry = qk_target_entry_new(QkGate_CX);
+///    qk_target_entry_add_property(cx_entry, qargs, 2, 0.0, 0.0);
+///    qk_target_add_instruction(target, cx_entry);
+///
+///    QkCircuit *circuit = qk_circuit_new(2, 0);
+///    qk_circuit_gate(circuit, QkGate_CX, (uint32_t[]){1,0}, NULL);
+///
+///    bool direction_ok = qk_transpiler_pass_standalone_check_gate_direction(circuit, target);
+/// ```
+///
+/// # Safety
+///
+/// Behavior is undefined if ``circuit`` or ``target`` is not a valid, non-null pointer to a ``QkCircuit`` and ``QkTarget``.
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_transpiler_pass_standalone_check_gate_direction(
+    circuit: *const CircuitData,
+    target: *const Target,
+) -> bool {
+    // SAFETY: Per documentation, the pointer is non-null and aligned.
+    let circuit = unsafe { const_ptr_as_ref(circuit) };
+    let target = unsafe { const_ptr_as_ref(target) };
+
+    let dag = match DAGCircuit::from_circuit_data(circuit, false, None, None, None, None) {
+        Ok(dag) => dag,
+        Err(e) => panic!("{}", e),
```

Comment:
Calling `to_string` implicitly here still needs the Python GIL. You might want to go two different ways with how this is handled:
1. Add rust-native errors to the rust native parts of the pass; or
2. Replace the panic with a specific message which will delete any information linked to the error message.

I understand that there is little time to get this done but we don't want to run into issues here and end up with an obscure pyo3 error from C.

Reply:
This is just for the circuit -> dag conversion, every standalone C API function has this exact thing. For this specific example it's just a simple step is just to do:
 
```suggestion
        Err(e) => panic!("Circuit to DAG conversion failed"),
```

Or even simpler is just it to use `.expect()` like in: https://github.com/Qiskit/qiskit/blob/main/crates/cext/src/transpiler/passes/remove_diagonal_gates_before_measure.rs#L49C18-L49C52

We just need to be a bit tactical about error returns and panics to C, what do we expect to be user actionable vs not, and what can be triggered from C or not. For example, a lot of `PyResult` returns they come from a control flow path only and in the case of no control flow there is no `PyErr` possible.

Reply:
Switched to using `expect` in the dag <> circuit conversions in https://github.com/Qiskit/qiskit/pull/14768/commits/b3cb14caca9df7847b9a8c0f35973bc4cab1c054